pyccel/ast/core.py

```diff
@@ -204,6 +204,12 @@ def broadcast(shape_1, shape_2):
             new_shape.append(e2)
         elif e2 == 1:
             new_shape.append(e1)
+        elif isinstance(e1, PyccelArraySize) and isinstance(e2, PyccelArraySize):
+            new_shape.append(e1)
+        elif isinstance(e1, PyccelArraySize):
+            new_shape.append(e2)
+        elif isinstance(e2, PyccelArraySize):
+            new_shape.append(e1)
```

Comment:
Should warnings be raised here?

Reply:
I am not sure. Can you give me a Python example where we enter in one of these cases?

Reply:
This actually occurs in lots of places, for example in most of our array tests in epyccel/test_arrays.
E.g:
```
@types( 'int32[:]', 'int32[:]' )
def array_int32_1d_add( x, y ):
    x[:] += y
```
For this calculation we are assuming that x and y have the same size. Up till now this has been done by setting the size to `None` with the modification above this would be explicit. I think we will need it to be explicit in order to carry out this operation in C. In addition once this assumption is explicit then we know the size of c in the following function:
```
@types('int[:]','int[:]')
def print_add(a, b):
    c = a + b
    print(c)
```

Reply:
I am not sure what should happen if we pass the function arrays of different sizes though. In numpy this would raise a `ValueError`. A short test shows that calling the following (translated):
```
a = np.ones(3)
b = np.ones(4)
print_add(a,b)
```
gives
```
2.0, 2.0, 2.0
```
and the following:
```
a = np.ones(4)
b = np.ones(3)
print_add(a,b)
```
gives
```
2.0, 2.0, 2.0, 1.0
```

Reply:
We don't get any warnings from the Fortran compiler?

Reply:
The fortran compiler has the same problem we do: it doesn't know the size of the arrays that will be passed to the function. There is nothing inherently wrong with the function:
```
!........................................
subroutine test_add(a, b)

  implicit none

  real(kind=8), intent(in) :: a(0:)
  real(kind=8), intent(in) :: b(0:)
  real(kind=8), allocatable :: c(:)

  allocate(c(0:size(a, 1) - 1_8))
  c = a + b
  print *, c

end subroutine test_add
!........................................
```
And I don't think it looks at the contents of the function when compiling the call

Reply:
Could you paste the generated Fortran code to this discussion?

Reply:
Even with `-Wall` I don't get any errors and the code runs

Reply:
I see, then we have a potential memory bug. Try compiling with `-fcheck=all` and run the code again...

Reply:
With `-fcheck=all` I have a runtime error:
```
Fortran runtime error: Array bound mismatch for dimension 1 of array 'b' (3/4)
```

Reply:
I am not sure how we should handle this case. A very complicated but effective method would be to create "assumptions" for the arguments. These assumptions could then be checked during the translation or in the wrapper (when we no longer use f2py, or in the f2py_XX.f90 file).
E.g. for:
```
@types('int[:]','int[:]')
def print_add(a, b):
    c = a + b
    print(c)
```
We would generate the assumption `shape(a,1)==shape(b,1)`

If this is not true when we call the function then we can raise an error. In the cwrapper it is possible to raise python errors e.g. `ValueError` as numpy does

This is beyond the scope of this pull request though. Currently this PR reproduces the same behaviour as before it simply makes it explicit

Reply:
This makes a lot of sense in Fortran: one should run the test suite in debug mode, where the code is compiled with lots of compile- and runtime-checks. And then recompile it in release mode (with lots of optimization and few checks) for the production runs.

Reply:
So I think we can be happy of this fact: where Numpy gives a runtime error we also get a runtime error in Fortran, provided that we add the proper compilation flags to the compiler.

Reply:
If we implement `assert` then you're solution is also possible. So for the following codes:
**Code A**
```
import numpy as np

@types('real[:]','real[:]')
def test_add(a,b):
    c = a + b
    print(c)

x = np.ones(3)
y = np.ones(4)

test_add(a,b)
```

**Code B**
```
import numpy as np

@types('real[:]','real[:]')
def test_add(a,b):
    c = a + b
    print(c)
```
we would generate an assumption `shape(a,1)==shape(b,1)` for the function `test_add`. We then have 2 options:
1. Code A generates a pyccel error: `"The function test_add is called with incompatible arguments"`. Code B compiles but the wrapper will raise a `ValueError`
2. Code A and B generate:
```
module prog

implicit none

contains

!........................................
subroutine test_add(a, b)

  implicit none

  real(kind=8), intent(in) :: a(0:)
  real(kind=8), intent(in) :: b(0:)
  real(kind=8), allocatable :: c(:)

  ! TODO: How to disable this check in release?
  if (shape(a,1) != shape(a,2)) then stop 'Assertion Error'

  allocate(c(0:size(a, 1) - 1_8))
  c = a + b
  print *, c

end subroutine test_add
!........................................

end module prog
```

I suspect the best answer is a hybrid where we do option 2 but we still check the assumptions in the cwrapper to raise python errors rather than having a Fortran runtime error which could cause problems for an ipython instance for example.
If you are happy with this method then I can create an issue which can be tackled when someone has lots of free time

Reply:
We already add the necessary flag (https://github.com/pyccel/pyccel/blob/master/pyccel/codegen/utilities.py#L59):
```
        if debug:
            flags += " -fbounds-check"
```

Reply:
> ```python
>         if debug:
>             flags += " -fbounds-check"
> ```

I think that's an old deprecated flag. The new way to do this is `-fcheck=bounds`.

Reply:
> I suspect the best answer is a hybrid where we do option 2 but we still check the assumptions in the cwrapper to raise python errors rather than having a Fortran runtime error which could cause problems for an ipython instance for example.
> If you are happy with this method then I can create an issue which can be tackled when someone has lots of free time

Yes, I think that makes sense. But a deeper discussion may be needed. Please create a dedicated issue where you can copy-paste from this thread as much as you want.